File: 01. Arrays/01. 2Sum Problem/main.cpp
#include <bits/stdc++.h>
using namespace std;

// Brute Force Approach: O(N^2) Time, O(1) Space
vector<int> twoSumBruteForce(vector<int>& nums, int target) {
    int n = nums.size();
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if (nums[i] + nums[j] == target)
                return {nums[i], nums[j]}; // Return indices of the two numbers
        }
    }
    return {}; // No solution found
}

// Sorting + Two Pointers Approach: O(N log N) Time, O(1) Space
vector<int> twoSumTwoPointers(vector<int> nums, int target) {
    vector<pair<int, int>> indexedNums;
    for (int i = 0; i < nums.size(); i++)
        indexedNums.push_back({nums[i], i}); // Store value with original index

    sort(indexedNums.begin(), indexedNums.end()); // Sort based on values

    int left = 0, right = nums.size() - 1;
    while (left < right) {
        int sum = indexedNums[left].first + indexedNums[right].first;
        if (sum == target)
            return {indexedNums[left].first, indexedNums[right].first}; // Return original indices
        else if (sum < target)
            left++;
        else
            right--;
    }
    return {}; // No solution found
}

// Hashing (Optimal) Approach: O(N) Time, O(N) Space
vector<int> twoSumHashing(vector<int>& nums, int target) {
    unordered_map<int, int> numMap; // Stores {value, index}
    
    for (int i = 0; i < nums.size(); i++) {
        int complement = target - nums[i];
        if (numMap.find(complement) != numMap.end())
            return {complement, nums[i]}; // Return indices of the two numbers
        numMap[nums[i]] = i; // Store index of the current number
    }
    return {}; // No solution found
}

int main() {
    vector<int> nums = {2, 7, 11, 15};
    int target = 9;

    // Brute Force Approach
    vector<int> result1 = twoSumBruteForce(nums, target);
    if (!result1.empty())
        cout << "Brute Force: [" << result1[0] << ", " << result1[1] << "]\n";
    else
        cout << "No solution found using Brute Force\n";

    // Two Pointers Approach
    vector<int> result2 = twoSumTwoPointers(nums, target);
    if (!result2.empty())
        cout << "Two Pointers: [" << result2[0] << ", " << result2[1] << "]\n";
    else
        cout << "No solution found using Two Pointers\n";

    // Hashing Approach
    vector<int> result3 = twoSumHashing(nums, target);
    if (!result3.empty())
        cout << "Hashing: [" << result3[0] << ", " << result3[1] << "]\n";
    else
        cout << "No solution found using Hashing\n";

    return 0;
}

File: 01. Arrays/01. 2Sum Problem/main.go
package main

import (
	"fmt"
	"sort"
)

// Brute Force Approach: O(N^2) Time, O(1) Space
func twoSumBruteForce(nums []int, target int) []int {
	n := len(nums)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			if nums[i]+nums[j] == target {
				return []int{nums[i], nums[j]} // Return values of the two numbers
			}
		}
	}
	return nil // No solution found
}

// Sorting + Two Pointers Approach: O(N log N) Time, O(N) Space (for sorting + indexing)
func twoSumTwoPointers(nums []int, target int) []int {
	type pair struct {
		value int
		index int
	}

	indexedNums := make([]pair, len(nums))
	for i, v := range nums {
		indexedNums[i] = pair{v, i}
	}

	sort.Slice(indexedNums, func(i, j int) bool {
		return indexedNums[i].value < indexedNums[j].value
	})

	left, right := 0, len(nums)-1
	for left < right {
		sum := indexedNums[left].value + indexedNums[right].value
		if sum == target {
			return []int{indexedNums[left].value, indexedNums[right].value} // Return values
		} else if sum < target {
			left++
		} else {
			right--
		}
	}
	return nil // No solution found
}

// Hashing (Optimal) Approach: O(N) Time, O(N) Space
func twoSumHashing(nums []int, target int) []int {
	numMap := make(map[int]int) // value -> index

	for _, num := range nums {
		complement := target - num
		if _, found := numMap[complement]; found {
			return []int{complement, num} // Return values
		}
		numMap[num] = 1
	}
	return nil // No solution found
}

func main() {
	nums := []int{2, 7, 11, 15}
	target := 9

	result1 := twoSumBruteForce(nums, target)
	if result1 != nil {
		fmt.Printf("Brute Force: [%d, %d]\n", result1[0], result1[1])
	} else {
		fmt.Println("No solution found using Brute Force")
	}

	result2 := twoSumTwoPointers(nums, target)
	if result2 != nil {
		fmt.Printf("Two Pointers: [%d, %d]\n", result2[0], result2[1])
	} else {
		fmt.Println("No solution found using Two Pointers")
	}

	result3 := twoSumHashing(nums, target)
	if result3 != nil {
		fmt.Printf("Hashing: [%d, %d]\n", result3[0], result3[1])
	} else {
		fmt.Println("No solution found using Hashing")
	}
}
